<?xml version="1.0"?>

<!--
    Copyright 2021 U.S. Department of Transportation, Federal Highway Administration

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
-->

<launch>
    <!-- VESC wrapper that interfaces with CARMA -->
    <node name="c1t_vesc_driver" pkg="c1t_vesc_driver" type="c1t_vesc_driver_node" output="screen">
        <remap from="vehicle_cmd" to="controller/vehicle_cmd"/>
        <remap from="enable_robotic" to="controller/enable_robotic"/>
        <remap from="robot_status" to="controller/robot_status"/>
        <remap from="ackermann_cmd" to="controller/ackermann_cmd"/>
        <remap from="sensors/core" to="controller/sensors/core"/>
        <rosparam command="load" file="$(find c1t_vesc_driver)/config/parameters.yaml"/>
    </node>

    <group ns="controller">
        <!-- Param file includes parameters for both message converter and VESC driver nodes -->
        <arg name="vesc_config" default="$(find c1t_vesc_driver)/config/vesc.yaml"/>
        <rosparam file="$(arg vesc_config)" command="load"/>

        <!-- Converts AckermannDriveStamped messages into corresponding motor and servo commands -->
        <node name="ackermann_to_vesc" pkg="vesc_ackermann" type="ackermann_to_vesc_node" output="screen"/>

        <!-- Actual VESC driver -->
        <node name="vesc_driver" pkg="vesc_driver" type="vesc_driver_node" output="screen">
            <param name="port" value="/dev/vesc"/>
        </node>
    </group>
</launch>
